src/slack/SlackControlManager.ts

```diff
@@ -1,67 +1,233 @@
 
-/**
- * This won't be available until they've connected to spotify
- */
-export async function connectSlack() {
-    const jwt = getItem("jwt");
-    const encodedJwt = encodeURIComponent(jwt);
-    const qryStr = `integrate=slack&plugin=musictime&token=${encodedJwt}`;
-
-    // authorize the user for slack
-    const endpoint = `${api_endpoint}/auth/slack?${qryStr}`;
-    launchWebUrl(endpoint);
-    refetchSlackConnectStatusLazily();
+const queryString = require("query-string");
+const { WebClient } = require("@slack/web-api");
+
+export async function connectSlackWorkspace() {
+  const qryStr = queryString.stringify({
+    plugin: getPluginType(),
+    plugin_uuid: getPluginUuid(),
+    pluginVersion: getVersion(),
+    plugin_id: getPluginId(),
+    auth_callback_state: getAuthCallbackState(),
+    integrate: "slack",
+  });
+
+  const url = `${api_endpoint}/auth/slack?${qryStr}`;
+
+  // authorize the user for slack
+  launchWebUrl(url);
+  // lazily check if the user has completed the slack authentication
+  setTimeout(() => {
+    refetchSlackConnectStatusLazily(40);
+  }, 10000);
 }
 
-export async function showSlackChannelMenu() {
-    let menuOptions = {
-        items: [],
-        placeholder: "Select a channel",
-    };
-
-    // get the available channels
-    const channelNames = await getChannelNames();
-    channelNames.sort();
-
-    channelNames.forEach((channelName) => {
-        menuOptions.items.push({
-            label: channelName,
-        });
-    });
+export async function disconnectSlack() {
+  const workspaces = getSlackWorkspaces();
+  if (workspaces.length === 0) {
+    window.showErrorMessage("Unable to find Slack integration to disconnect");
+    return;
+  }
+
+  // show a selection of which one or all workspaces to disconnect
+  const selectedItem = await showSlackWorkspacesToDisconnect();
+  if (selectedItem) {
+    const authId = selectedItem.value;
+    const domain = selectedItem.label;
+    let msg = "";
+    if (authId === "all") {
+      msg = "Are you sure you would like to disconnect all Slack workspaces?";
+    } else {
+      msg = `Are you sure you would like to disconnect the '${domain}' Slack workspace?`;
+    }
+
+    // ask before disconnecting
+    const selection = await window.showInformationMessage(msg, ...[DISCONNECT_LABEL]);
 
-    const pick = await showQuickPick(menuOptions);
-    if (pick && pick.label) {
-        return pick.label;
+    if (selection === DISCONNECT_LABEL) {
+      if (authId === "all") {
+        for await (const workspace of workspaces) {
+            await softwarePut(`/auth/slack/disconnect`, { authId: workspace.authId }, getItem("jwt"));
+            removeSlackIntegration(workspace.authId);
+        }
+        window.showInformationMessage("Disconnected selected Slack integrations");
+      } else {
+        await softwarePut(`/auth/slack/disconnect`, { authId }, getItem("jwt"));
+        removeSlackIntegration(authId);
```

Comment:
will the local integrations ever become out of sync with the server integrations if the put request fails?

Reply:
thats a good question and I thought about that but didn't have a clean way of handling this yet